File: flask_app/validation.py
import re
import os

ALLOWED_EXTENSIONS = {"jpg", "jpeg", "png", "gif", "mp4", "mov"}
ALLOWED_MIMETYPES  = {"image/jpeg", "image/png", "image/gif", "video/mp4", "video/quicktime"}

TIPOS_MIEMBRO    = {"pregrado", "postgrado", "funcionario", "academico"}
TIPOS_ACTIVIDAD  = {"deportiva", "artistica", "tecnologica", "social", "recreativa"}
DIAS_SEMANA      = {"lunes", "martes", "miercoles", "jueves", "viernes", "sabado", "domingo"}


def validar_nombre(valor):
    return len(valor.strip()) >= 3


def validar_apellido(valor):
    return len(valor.strip()) >= 3


def validar_email(valor):
    patron = r'^[^\s@]+@[^\s@]+\.[^\s@]+$'
    return bool(re.match(patron, valor.strip()))


def validar_telefono(valor):
    if not valor or valor.strip() == "":
        return True
    patron = r'^\+?[0-9]{7,15}$'
    return bool(re.match(patron, valor.strip()))


def validar_tipo_miembro(valor):
    return valor in TIPOS_MIEMBRO


def validar_tipo_actividad(valor):
    return valor in TIPOS_ACTIVIDAD


def validar_nombre_actividad(valor):
    return len(valor.strip()) > 0


def validar_dia(valor):
    return valor in DIAS_SEMANA


def validar_hora(valor):
    if not isinstance(valor, str):
        return False
    patron = r'^([01]\d|2[0-3]):[0-5]\d$'
    return bool(re.match(patron, valor.strip()))


def validar_horas(hora_inicio, hora_fin):
    if not validar_hora(hora_inicio) or not validar_hora(hora_fin):
        return False
    return hora_fin > hora_inicio


def validar_enlace(valor):
    if not isinstance(valor, str):
        return False
    patron = r'^https?://.+\..+'
    return bool(re.match(patron, valor.strip()))


def validar_archivo(archivo):
    if archivo is None or archivo.filename == "":
        return False

    extension = archivo.filename.rsplit(".", 1)[-1].lower() if "." in archivo.filename else ""
    if extension not in ALLOWED_EXTENSIONS:
        return False

    mimetype = archivo.mimetype
    if mimetype not in ALLOWED_MIMETYPES:
        return False

    return True

def validar_formulario(form, archivo):
    errores = {}

    if not validar_nombre(form.get("nombre", "")):
        errores["nombre"] = "El nombre debe tener al menos 3 letras."

    if not validar_apellido(form.get("apellido", "")):
        errores["apellido"] = "El apellido debe tener al menos 3 letras."

    if not validar_email(form.get("email", "")):
        errores["email"] = "Ingrese un correo electrónico válido."

    if not validar_telefono(form.get("telefono", "")):
        errores["telefono"] = "Ingrese un número de teléfono válido. Ej: +56912345678"

    if not validar_tipo_miembro(form.get("tipo-miembro", "")):
        errores["tipo-miembro"] = "Seleccione un tipo de miembro válido."

    if not validar_tipo_actividad(form.get("tipo-actividad", "")):
        errores["tipo-actividad"] = "Seleccione un tipo de actividad válido."
    print("antes de validar nombre actividad")
    if not validar_nombre_actividad(form.get("nombre-actividad", "")):
        errores["nombre-actividad"] = "Ingrese el nombre de la actividad."
    print("después de validar nombre actividad, errores:", errores)

    dias_marcados = form.getlist("dias")
    if not dias_marcados:
        errores["dias"] = "Seleccione al menos un día."
    else:
        for dia in dias_marcados:
            if not validar_dia(dia):
                errores["dias"] = "Día inválido."
                break
            hora_inicio = form.get(f"inicio-{dia}", "")
            hora_fin    = form.get(f"fin-{dia}", "")
            if not validar_horas(hora_inicio, hora_fin):
                errores[f"horas-{dia}"] = "Ingrese horas válidas. La hora de término debe ser posterior a la de inicio."

    if not validar_enlace(form.get("enlace-actividad", "")):
        errores["enlace-actividad"] = "Ingrese un enlace válido. Ej: https://instagram.com/mi_club"

    if not validar_archivo(archivo):
        errores["archivo-actividad"] = "Adjunte al menos una foto o video válido (JPG, PNG, GIF, MP4, MOV)."

    return errores


def validar_comentario(nombre, texto):
    errores = {}

    nombre = nombre.strip() if isinstance(nombre, str) else ""
    texto = texto.strip() if isinstance(texto, str) else ""

    if len(nombre) < 3 or len(nombre) > 80:
        errores["nombre"] = "El nombre debe tener entre 3 y 80 caracteres."

    if len(texto) < 5:
        errores["texto"] = "El comentario debe tener al menos 5 caracteres."
    elif len(texto) > 300:
        errores["texto"] = "El comentario no puede superar 300 caracteres."

    return errores

File: flask_app/static/js/validation.js
const validarNombre = () => {
  const input = document.getElementById("nombre");
  const error = document.getElementById("error-nombre");
  const valor = input.value.trim();
 
  if (valor.length < 3) {
    error.innerText = "El nombre debe tener al menos 3 letras.";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}
 
const validarApellido = () => {
  const input = document.getElementById("apellido");
  const error = document.getElementById("error-apellido");
  const valor = input.value.trim();
 
  if (valor.length < 3) {
    error.innerText = "El apellido debe tener al menos 3 letras.";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}
 
const validarCorreo = () => {
  const input = document.getElementById("email");
  const error = document.getElementById("error-email");
  const valor = input.value.trim();
  const formatoCorreo = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
 
  if (!formatoCorreo.test(valor)) {
    error.innerText = "Ingrese un correo electrónico válido. Ej: tomas.raggio@ug.uchile.cl";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}
 
const validarTelefono = () =>  {
  const input = document.getElementById("telefono");
  const error = document.getElementById("error-telefono");
  const valor = input.value.trim();
  const formatoTelefono = /^\+?[0-9]{7,15}$/;
 
  if (valor === "") {
    error.innerText = "";
    input.classList.remove("input-error");
    return true;
  }
 
  if (!formatoTelefono.test(valor)) {
    error.innerText = "Ingrese un número de teléfono válido. Ej: +56912345678";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}
 
const validarTipoMiembro = () => {
  const input = document.getElementById("tipo-miembro");
  const error = document.getElementById("error-tipo-miembro");
 
  if (input.value === "") {
    error.innerText = "Seleccione un tipo de miembro.";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}

const validarRegion = () => {
  const input = document.getElementById("region");
  const error = document.getElementById("error-region");

  if (input.value === "") {
    error.innerText = "Seleccione una region.";
    input.classList.add("input-error");
    return false;
  }

  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}

const validarComuna = () => {
  const input = document.getElementById("comuna");
  const error = document.getElementById("error-comuna");

  if (input.value === "") {
    error.innerText = "Seleccione una comuna.";
    input.classList.add("input-error");
    return false;
  }

  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}

const validarTipoActividad = () => {
  const input = document.getElementById("tipo-actividad");
  const error = document.getElementById("error-tipo-actividad");
 
  if (input.value === "") {
    error.innerText = "Seleccione un tipo de actividad.";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}
 

 
const validarDias = () => {
  const checkboxes = document.querySelectorAll("input[name='dias']:checked");
  const error = document.getElementById("error-dias");
 
  if (checkboxes.length === 0) {
    error.innerText = "Seleccione al menos un día.";
    return false;
  }
 
  error.innerText = "";
  return true;
}
 
const validarHorasDia = (dia) => {
  const inicio = document.getElementById("inicio-" + dia);
  const fin = document.getElementById("fin-" + dia);
  const errorHora = document.getElementById("error-hora-" + dia);
  let valido = true;
  console.log("inicio:", inicio.value, "fin:", fin.value);
  if ((inicio.value === "") || (fin.value === "")){
    errorHora.innerText = "Ingrese una hora válida";
    valido = false;
  } else if (fin.value <= inicio.value) {
  errorHora.innerText = "La hora de término debe ser posterior a la de inicio.";
  valido = false; 
    
  } else {
    errorHora.innerText = "";
    fin.classList.remove("input-error");
  }
 
  return valido;
}
 
const validarHorasActividad = () => {
  const dias = ["lunes", "martes", "miercoles", "jueves", "viernes", "sabado", "domingo"];
  let valido = true;
 
  dias.forEach((dia) => {
    const checkbox = document.querySelector("input[name='dias'][value='"+dia+"']");
    if (checkbox && checkbox.checked) {
      if (!validarHorasDia(dia)) {
        valido = false;
      }
    }
  });
 
  return valido;
}
 
const validarArchivo = () => {
  const input = document.getElementById("archivo-actividad");
  const error = document.getElementById("error-archivo");
 
  if (input.files.length === 0) {
    error.innerText = "Adjunte al menos una foto o video de la actividad.";
    return false;
  }
 
  error.innerText = "";
  return true;
}
 
const validarEnlace = () => {
  const input = document.getElementById("enlace-actividad");
  const error = document.getElementById("error-enlace");
  const valor = input.value.trim();
  const formatoUrl = /^https?:\/\/.+\..+/;
 
  if (!formatoUrl.test(valor)) {
    error.innerText = "Ingrese un enlace válido. Ej: https://instagram.com/aaaa";
    input.classList.add("input-error");
    return false;
  }
 
  error.innerText = "";
  input.classList.remove("input-error");
  return true;
}

const validarFormulario = () => {
  let valido = true;
 
  if (!validarNombre())          valido = false;
  if (!validarApellido())        valido = false;
  if (!validarCorreo())          valido = false;
  if (!validarTelefono())        valido = false;
  if (!validarTipoMiembro())     valido = false;
  if (!validarRegion())          valido = false;
  if (!validarComuna())          valido = false;
  if (!validarTipoActividad())   valido = false;
  if (!validarDias())            valido = false;
  if (!validarHorasActividad())  valido = false;
  if (!validarArchivo())         valido = false;
  if (!validarEnlace())          valido = false;
 
  return valido;
}

const agregarOpcion = (select, valor, texto) => {
  const opcion = document.createElement("option");
  opcion.value = valor;
  opcion.innerText = texto;
  select.appendChild(opcion);
}

const limpiarSelectComuna = (texto) => {
  const comuna = document.getElementById("comuna");
  comuna.innerHTML = "";
  agregarOpcion(comuna, "", texto);
  comuna.disabled = true;
}

const cargarComunas = async (regionId, comunaSeleccionada = "") => {
  limpiarSelectComuna("-- Cargando comunas --");

  if (regionId === "") {
    limpiarSelectComuna("-- Seleccione una region --");
    return;
  }

  try {
    const respuesta = await fetch("/api/regiones/" + regionId + "/comunas");
    const datos = await respuesta.json();

    if (!respuesta.ok || !datos.ok) {
      throw new Error("No fue posible cargar las comunas.");
    }

    const comuna = document.getElementById("comuna");
    comuna.innerHTML = "";
    agregarOpcion(comuna, "", "-- Seleccione --");

    datos.comunas.forEach((item) => {
      agregarOpcion(comuna, item.id, item.nombre);
    });

    comuna.disabled = false;

    if (comunaSeleccionada !== "") {
      comuna.value = comunaSeleccionada;
    }
  } catch (error) {
    limpiarSelectComuna("-- No fue posible cargar comunas --");
  }
}

const cargarRegiones = async () => {
  const form = document.getElementById("form-datos");
  const region = document.getElementById("region");
  const regionSeleccionada = form.dataset.regionSeleccionada;
  const comunaSeleccionada = form.dataset.comunaSeleccionada;

  try {
    const respuesta = await fetch("/api/regiones");
    const datos = await respuesta.json();

    if (!respuesta.ok || !datos.ok) {
      throw new Error("No fue posible cargar las regiones.");
    }

    datos.regiones.forEach((item) => {
      agregarOpcion(region, item.id, item.nombre);
    });

    if (regionSeleccionada !== "") {
      region.value = regionSeleccionada;
      await cargarComunas(regionSeleccionada, comunaSeleccionada);
    }
  } catch (error) {
    const errorRegion = document.getElementById("error-region");
    errorRegion.innerText = "No fue posible cargar las regiones.";
  }
}

const checkboxes = document.querySelectorAll("input[name='dias']");
 
checkboxes.forEach((checkbox) => {
  checkbox.addEventListener("change", () => {
    const contenedorHoras = document.getElementById("hora-" + checkbox.value);
    if (checkbox.checked) {
      contenedorHoras.classList.add("visible");
    } else {
      contenedorHoras.classList.remove("visible");
    }
  });
});

document.getElementById("region").addEventListener("change", (event) => {
  document.getElementById("error-region").innerText = "";
  document.getElementById("error-comuna").innerText = "";
  cargarComunas(event.target.value);
});

document.getElementById("btn-enviar").addEventListener("click", (event) => {
  console.log("click recibido");
  const vali = validarFormulario()
  console.log("validación:", vali);
  if (vali) {
    document.getElementById("form-datos").submit();
  };
  
});

cargarRegiones();
